### PR TITLE
Skip IPAddr validation for "host-gateway" string

### PR DESCRIPTION
**Pull request by arkodg**
Relates to - https://github.com/moby/moby/pull/40007
The above PR added support in moby, that detects if
a special string "host-gateway" is added to the IP
section of --add-host, and if true, replaces it with
a special IP value (value of --host-gateway-ip Daemon flag
which defaults to the IP of the default bridge).

This PR is needed to skip the validation for the above
feature

Signed-off-by: Arko Dasgupta <arko.dasgupta@docker.com>

### CODE REVIEW

**Comment by silvin-lubecki** [comment]
Related to #2290

File: opts/hosts.go [hunk 1]
```diff
@@ -24,6 +24,9 @@ var (
 	DefaultTLSHost = fmt.Sprintf("tcp://%s:%d", DefaultHTTPHost, DefaultTLSHTTPPort)
 	// DefaultNamedPipe defines the default named pipe used by docker on Windows
 	DefaultNamedPipe = `//./pipe/docker_engine`
+	// HostGatewayName defines a special string which users can append to --add-host
+	// to add an extra entry in /etc/hosts that maps host.docker.internal to the host IP
+	HostGatewayName = "host-gateway"
```
Comment (thaJeztah) [inline opts/hosts.go]: This should be a `const`; also wondering if there's a const for this in moby/moby that we can use

File: opts/hosts.go [hunk 2]
```diff
@@ -160,8 +163,11 @@ func ValidateExtraHost(val string) (string, error) {
 	if len(arr) != 2 || len(arr[0]) == 0 {
 		return "", fmt.Errorf("bad format for add-host: %q", val)
 	}
-	if _, err := ValidateIPAddress(arr[1]); err != nil {
-		return "", fmt.Errorf("invalid IP address in add-host: %q", arr[1])
+	// Skip IPaddr validation for "host-gateway" string
+	if arr[1] != HostGatewayName {
+		if _, err := ValidateIPAddress(arr[1]); err != nil {
```
Comment (thaJeztah) [inline opts/hosts.go]: Actually slightly on the fence if we should validate _at all_ in the cli, or if we should just pass it to the daemon and have the daemon validate / produce an error (haven't checked yet what the UX looks like in that case)

File: opts/hosts.go (hunk 1 above)
Comment (silvin-lubecki) [inline opts/hosts.go]: I think this value should be used directly from [moby](https://github.com/moby/moby/pull/40007/files#diff-65c8668a717635f8e126e71c7213f537R7)

**Comment by thaJeztah** [comment]
Tested what it looks like if we remove the IP address validation;

```bash
docker create --add-host foo:fooobar nginx:alpine
Error response from daemon: invalid IP address in add-host: "fooobar"

docker run --add-host foo:fooobar nginx:alpine
docker: Error response from daemon: invalid IP address in add-host: "fooobar".
```

And, I think those errors look reasonable (have not tried `docker stack deploy` though)

**Comment by thaJeztah** [comment]
`docker service create` also validates on the service, but appears to be retrying creation; perhaps it's only validating the moment the container is created and not when the _service_ is created;

```bash
docker service create --host foo:foooobar nginx:alpine
opbff21havwdz5fyrjk68so11
overall progress: 0 out of 1 tasks
1/1: invalid IP address in add-host: "foooobar"

...

overall progress: 0 out of 1 tasks
1/1: assigned  [======================>                            ]

...

overall progress: 0 out of 1 tasks
1/1: invalid IP address in add-host: "foooobar"
```

**Comment by arkodg** [comment]
sure, there is already an identical validation in the daemon, and this PR is just here to unblock this feature.
we can completely remove `ValidateHost` from the `cli`, but I'm not sure what the design philosophy is for cli and daemon arg validation

**Comment by thaJeztah** [comment]
Overall, I prefer delegating validation to the daemon as much as possible; at _most_ validating well-formedness of values (which in this case was possible before).

Keeping the daemon as source of truth not only keeps complexity out of the cli code, it also prevents situations where the cli talks to a different (older/newer) version of the daemon, or (eg) a daemon running on a different platform.

We should probably (as a follow up) look at the `docker service create` example though, because that UX isn't ideal (if possible, validate earlier, or perhaps give up if the config is invalid instead of keep retrying)

**Comment by arkodg** [comment]
@thaJeztah IMHO for the `docker service create` case, I think keeping the logic here makes sense, updated the PR removing the `const`

**Comment by silvin-lubecki** [comment]
@arkodg The vendoring looks weird:
```
These files were changed:

?? vendor/github.com/opencontainers/runc/libcontainer/cgroups/
?? vendor/github.com/opencontainers/runc/libcontainer/configs/
```

**Comment by silvin-lubecki** [comment]
Please also rebase 🦁

**Comment by thaJeztah** [comment]
@arkodg could you;

- move the vendor update to a separate commit (first commit)
    - only including changes needed related to that update (I see some tests failing, which could be related to changes upstream)
- move local changes to for this feature to a commit coming after that (combine the local changes in the current commits)

**Comment by arkodg** [comment]
@thaJeztah the issue is with `make cross` using the latest `docker/docker` reappoint

**Comment by thaJeztah** [comment]
opened https://github.com/docker/cli/pull/2333 to update vendoring

**Comment by thaJeztah** [comment]
@arkodg I think this was a bad rebase, as I see no local code-changes anymore, other than adding the import 🤔 

Noticed this when I rebased the PR (also the commit message was now only about `vndr`); see https://github.com/docker/cli/compare/master...thaJeztah:carry_2293_add_host_gateway_dns_entry?expand=1

**Comment by thaJeztah** [comment]
@silvin-lubecki found an old commit here; https://github.com/docker/cli/commit/709cf07122c91e111c53edd4457e48fd87777fb1

Updating my branch with that

**Comment by arkodg** [comment]
@thaJeztah I removed the old commit from the PR after you suggested to rc the CI issue by separating the commits to a vendor commit and the feature commit :)

**Comment by yuraku** [comment]
Hi guys. Is there any ETA on when this PR will be merged? 🤞🏻

**Pull request closed without merging**